doc/config.md
## Confy: Configuration
To change these options, just add `cfg.theVariable = value` at the top of your `build.nim` file.  
*Note: Refer to @[confy/cfg.nim](./src/confy/cfg.nim) if something is not behaving quite right. This list might not be fully in sync.*
```nim
import confy
```
```nim
# General
cfg.verbose      = on           ## Makes the cli output information completely verbose. (for debugging)
cfg.quiet        = on           ## Makes the cli output information to be as minimal as possible.  (for cleaner cli output)  (default: on)
cfg.fakeRun      = off          ## Everything will run normally, but commands will not really be executed if set to `on`.
cfg.cores        = 0.8 * cores  ## Total cores to use for compiling.  (default = 80% of max)
cfg.prefix       = "confy: "    ## Prefix that will be added at the start of every command output.
cfg.tab          = "     : "    ## Tab that will be added at the start of every new line in of the same message.
```
```nim
# Folders
cfg.rootDir      = build/".."          ## Assumes the build.nim file output is stored inside root/bin/, so going back one gets to `REPO/*`
cfg.srcDir       = rootDir/"src"       ## Root Folder where source code files are searched for first (source code root)
cfg.binDir       = rootDir/"bin"       ## Root Folder where files will output (binary output root)
cfg.libDir       = rootDir/"lib"       ## Root Folder where libraries are stored
cfg.docDir       = rootDir/"doc"       ## Root Folder where the documentation files will go
cfg.examplesDir  = rootDir/"examples"  ## Root Folder where the examples for the project are stored
cfg.testsDir     = rootDir/"tests"     ## Root Folder that contains the test suite of the library/project
cfg.cacheDir     = binDir/".cache"     ## Subfolder where the compilation cache data is output
cfg.zigDir       = binDir/"zig"        ## Subfolder where the zig binaries 
```
```nim
# Files
cfg.file    = "build.nim"         ## File used for storing the builder config/app. Searched for @`srcDir/cfg.file`
cfg.db      = binDir/".confy.db"  ## File used for storing the builder database.
cfg.zigJson = binDir/".zig.json"  ## Zig download index json file. Will be output when the ZigCC binaries are downloaded.
```
```nim
# Commands
cfg.nimble  = &"nimble {switchVerbose}"              ## Command to run when nimble needs to be run
cfg.nimc    = &"nim c {switchVerbosity} -d:release"  ## Command to run for nimc tasks
```
```nim
# Compilation
## Flags  (see confy/flags.nim for details)
cfg.flagsC  = fl.allC           ## C   : Preset Flags object for all warnings and errors active
cfg.flags   = fl.allPP          ## C++ : Preset Flags object for all warnings and errors active

## Other options
cfg.Cstr         = "CC"         ## Prefix used for formatting the quiet output calls to the Compiler.
cfg.Lstr         = "LD"         ## Prefix used for formatting the quiet output calls to the Linker.
cfg.zigSystemBin = on           ## Uses the System's ZigCC path, without downloading a new version from the web.
```



src/confy/cfg.nim
#:_____________________________________________________
#  confy  |  Copyright (C) Ivan Mar (sOkam!)  |  MIT  :
#:_____________________________________________________
# std dependencies
import std/os except `/`
when not defined(nimscript):
  import std/paths
import std/strformat
import std/cpuinfo
# confy dependencies
import ./types
import ./auto
import ./flags as fl

#_______________________________________
# confy: Configuration defaults
#___________________
when defined(nimscript):
  var cores *:int= 1
else:
  var cores *:int= (0.8 * countProcessors().float).int
  ## Total cores to use for compiling.  (default = 80% of max)
var verbose *:Opt=     off
  ## Output will be fully verbose when active.
var quiet   *:Opt=     on
  ## Output will be formatted in a minimal clean style when active.
var prefix  *:string=  "confy: "
  ## Prefix that will be added at the start of every command output.
var tab     *:string=  "     : "
  ## Tab that will be added at the start of every new line in of the same message.
var Cstr    *:string=  "CC"
  ## Prefix used for formatting the quiet output calls to the Compiler.
var Lstr    *:string=  "LD"
  ## Prefix used for formatting the quiet output calls to the Linker.
var zigSystemBin * = on  # default:on
  ## Uses the System's ZigCC path, without downloading a new version from the web.
  ## When on : Uses the system's zig like `zig cc file.c
  ## When off: Runs the zig compiler setup logic and executes the zig compiler like `cfg.zigDir/zig cc file.c`


#_________________________________________________
# confy: Debugging
#___________________
var fakeRun  *:Opt=  off
  ## Everything will run normally, but commands will not really be executed.


#_____________________________
# Project: Folders
#___________________
var rootDir *:Dir=  Dir(".")  # Assumes the nimble/nims file is in root/, and is called from that folder.
when not defined(nimscript):
  rootDir = Dir(getAppDir()/"..")  # Assume the builder is inside root/bin/
# Root Folders
var srcDir       *:Dir=  rootDir/"src"
var binDir       *:Dir=  rootDir/"bin"
var libDir       *:Dir=  rootDir/"lib"
var docDir       *:Dir=  rootDir/"doc"
var examplesDir  *:Dir=  rootDir/"examples"
var testsDir     *:Dir=  rootDir/"tests"
#___________________
# Subfolders
var cacheDir     *:Dir=  binDir/".cache"
var zigDir       *:Dir=  binDir/"zig"

#_________________________________________________
# Project: Files
#___________________
var file    *:Fil=  "build.nim"
  ## File used for storing the builder config/app.
var db      *:Fil=  binDir/".confy.db"
  ## File used for storing the builder database.
var zigJson *:Fil=  binDir/".zig.json"
  ## Zig download index json file.

#_________________________________________________
# Nim: commands with Sane Defaults
#___________________
# Verbosity --
let switchVerbose   = if cfg.verbose: "--verbose" else: ""
let switchVerbosity = if cfg.verbose: "--verbosity:2" else: ""
# Commands
var nimble * = &"nimble {switchVerbose}"
var nimc   * = &"nim c {switchVerbosity} -d:release"


#_________________________________________________
# Compiler Configuration
#___________________
# Flags
var flagsC * = fl.allC
var flags  * = fl.allPP
